### ISSUE TITLE
Is it bad practice to use app.add_background_task() given it reuses the current connection

### ISSUE THREAD

**Issue by larsblumberg** [pgjones/quart-db#10]
[The database connection is stored in `g.connection` when auto-acquired in `QuartDB.before_request()`](https://github.com/pgjones/quart-db/blob/1573c9ef3d619867c2c683261c070c003fa0e3c4/src/quart_db/extension.py#L120).

Now [if `app.add_background_task()` is used the app context `g` is copied over to the background task](https://github.com/pallets/quart/blob/8c84efc3f208fb0d8339b2afd7a94b357bad2398/src/quart/app.py#L1564) so that

1. a single DB connection will be used in 2 different tasks, [which is not supported by the postgres protocol](https://github.com/MagicStack/asyncpg/issues/738#issuecomment-826149194)
2. [the connection will be prematurely released when the main task (the request handling function) ends](https://github.com/pgjones/quart-db/blob/1573c9ef3d619867c2c683261c070c003fa0e3c4/src/quart_db/extension.py#L124) while the background task still runs and it actually needs an open connection.

Above said, shouldn't `app.add_background_task()` be avoided so that the DB connection is not shared by two tasks?

**Comment by pgjones** [comment]
I agree this is a problem. I think usage of `g.connection` in a background task is a bad idea - I've not noticed this as I always use `async with quart_db.connection() as connection:` in a background task.

I think Quart is maybe at fault for copying rather than creating a new context for background tasks. Would you agree?

**Comment by larsblumberg** [comment]
> I think Quart is maybe at fault for copying rather than creating a new context for background tasks. Would you agree?

Yes, in this particular case of a non-sharable resources such as the db connection, this context attribute should not be copied. Then for the sake of framework simplicity, no context should be copied over to the background task at all. The framework still allows for context copying if a developer desires to do so.

PS: I just ran into a bug in my production system where it was slowly accumulating un-released database connections because of not releasing them back to the pgasync connection pool (not calling `QuartDB.release()`). Reason being the shared connection between a request and a background task it spawned. For some reason it happened that 2 connections have been acquired but only one was released in conjunction with a background task. I just stop using background tasks for now until I understand it better how I should be using background tasks properly in conjunction with connection acquiring.
